File: GCMarqueeViewDemo_OC/GCMarqueeViewDemo/GCMarqueeView/GCMarqueeItem.m
//
//  GCMarqueeItem.m
//  GCMarqueeView
//
//  Created by HenryCheng on 2019/8/1.
//  Copyright © 2019 HenryCheng. All rights reserved.
//

#import "GCMarqueeItem.h"
#import "GCMarqueeModel.h"

@interface GCMarqueeItem()

@property (nonatomic, strong) GCMarqueeModel *model;
@property (nonatomic, strong) UIImageView *imageV;
@property (nonatomic, strong) UILabel *titleLabel;

@end

static CGFloat titleFont = 12.0f;
static CGFloat padding = 18.0f;
static CGFloat icon_title_margin = 10.0f;
static CGFloat icon_width = 20.0f;
static CGFloat icon_height = 18.0f;

@implementation GCMarqueeItem

- (instancetype)initWithFrame:(CGRect)frame model:(GCMarqueeModel *)model {
    if (self == [super initWithFrame:frame]) {
        _model = model;
        self.layer.masksToBounds = YES;
        self.layer.cornerRadius = model.itemHeight / 2;
        self.backgroundColor = [UIColor orangeColor];
        [self initUI];
    }
    return self;
}

- (void)initUI {
    
    _imageV = [UIImageView new];
    _imageV.image = [UIImage imageNamed:@"flag_icon"];
    [self addSubview:_imageV];
    
    _titleLabel = [UILabel new];
    _titleLabel.font = [UIFont systemFontOfSize:titleFont];
    _titleLabel.text = _model.title;
    [_titleLabel sizeToFit];
    [self addSubview:_titleLabel];
    UITapGestureRecognizer *tapGesture = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(tap)];
    [self addGestureRecognizer:tapGesture];
}

- (void)layoutSubviews {
    [super layoutSubviews];
    _imageV.frame = CGRectMake(padding, (_model.itemHeight - icon_height) / 2, icon_width, icon_height);
    _titleLabel.frame = CGRectMake(padding + icon_width + icon_title_margin, (_model.itemHeight - 15) / 2, _model.itemWidth - padding * 2 - icon_width - icon_title_margin, 15);
}

- (void)tap {
    if (_block) {
        _block(_model);
    }
}

@end

File: GCMarqueeViewDemo_Swift/GCMarqueeViewDemo_Swift/GCMarqueeView/GCMarqueeItem.swift
//
//  GCMarqueeItem.swift
//  GCMarqueeViewDemo_Swift
//
//  Created by HenryCheng on 2019/8/2.
//  Copyright © 2019 igancao. All rights reserved.
//

import UIKit

class GCMarqueeItem: UIView {
    
    private let titleFont: CGFloat = 12.0;
    private let padding: CGFloat = 18.0;
    private let icon_title_margin: CGFloat = 10.0;
    private let defaultHeight: CGFloat = 28.0;
    private let icon_width: CGFloat = 20.0;
    private let icon_height: CGFloat = 18.0;

    var model: GCMarqueeModel?
    private var imageV: UIImageView?
    private var titleLabel: UILabel?

    init(frame: CGRect, model: GCMarqueeModel) {
        self.model = model
        super.init(frame: frame)
        self.layer.masksToBounds = true
        self.layer.cornerRadius = model.itemHeight / 2
        self.backgroundColor = .orange
        initUI()
    }
    
    override init(frame: CGRect) {
        super.init(frame: frame)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func layoutSubviews() {
        imageV!.frame = CGRect(x: padding, y: ((model?.itemHeight ?? defaultHeight) - icon_height) / 2, width: icon_width, height: icon_height);
        titleLabel!.frame = CGRect(x: padding + icon_width + icon_title_margin, y: ((model?.itemHeight ?? defaultHeight) - 15) / 2, width: model?.itemWidth ?? defaultHeight - padding * 2 - icon_width - icon_title_margin, height: 15)
    }
    
    func initUI() {
        imageV = UIImageView()
        imageV!.image = UIImage(named: "flag_icon")
        addSubview(imageV!)
        
        titleLabel = UILabel()
        titleLabel?.font = UIFont.systemFont(ofSize: titleFont)
        titleLabel?.text = model?.title
        addSubview(titleLabel!)
        
        let tapGuester = UITapGestureRecognizer(target: self, action: #selector(tap))
        addGestureRecognizer(tapGuester)
    }
    
    @objc func tap() {
        
    }
    

}
